presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MetadataQueryOptimizer.java

```diff
@@ -0,0 +1,192 @@
+    private final Metadata metadata;
+
+    public MetadataQueryOptimizer(Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
+    {
+        if (!isOptimizeMetadataQueries(session)) {
+            return Optional.empty();
+        }
+
+        if (!(node instanceof AggregationNode)) {
+            return Optional.empty();
+        }
+
+        AggregationNode aggregation = (AggregationNode) node;
+        // supported functions are only MIN/MAX/APPROX_DISTINCT or distinct aggregates
+        for (FunctionCall call : aggregation.getAggregations().values()) {
+            if (!ALLOWED_FUNCTIONS.contains(call.getName().toString()) && !call.isDistinct()) {
+                return Optional.empty();
+            }
+        }
+
+        Optional<TableScanNode> result = findTableScan(lookup.resolve(aggregation.getSource()), lookup);
+        if (!result.isPresent()) {
+            return Optional.empty();
+        }
+
+        // verify all outputs of table scan are partition keys
+        TableScanNode tableScan = result.get();
+
+        ImmutableMap.Builder<Symbol, Type> typesBuilder = ImmutableMap.builder();
+        ImmutableMap.Builder<Symbol, ColumnHandle> columnBuilder = ImmutableMap.builder();
+
+        List<Symbol> inputs = tableScan.getOutputSymbols();
+        for (Symbol symbol : inputs) {
+            ColumnHandle column = tableScan.getAssignments().get(symbol);
+            ColumnMetadata columnMetadata = metadata.getColumnMetadata(session, tableScan.getTable(), column);
+
+            typesBuilder.put(symbol, columnMetadata.getType());
+            columnBuilder.put(symbol, column);
+        }
+
+        Map<Symbol, ColumnHandle> columns = columnBuilder.build();
+        Map<Symbol, Type> types = typesBuilder.build();
+
+        // Materialize the list of partitions and replace the TableScan node
+        // with a Values node
+        TableLayout layout = null;
+        if (!tableScan.getLayout().isPresent()) {
+            List<TableLayoutResult> layouts = metadata.getLayouts(session, tableScan.getTable(), Constraint.alwaysTrue(), Optional.empty());
+            if (layouts.size() == 1) {
+                layout = Iterables.getOnlyElement(layouts).getLayout();
+            }
+        }
+        else {
+            layout = metadata.getLayout(session, tableScan.getLayout().get());
+        }
+        if (layout == null || !layout.getDiscretePredicates().isPresent()) {
+            return Optional.empty();
+        }
+        DiscretePredicates predicates = layout.getDiscretePredicates().get();
+
+        // the optimization is only valid if the aggregation node only relies on partition keys
+        if (!predicates.getColumns().containsAll(columns.values())) {
```

Comment:
I'm not sure I understand this bit.  Why isn't it
`if (!columns.values().containsAll(predicates.getColumns()))`